src/tracing.zig

```
@@ -0,0 +1,371 @@
+    }
+}
+
+pub fn deinitTracingTLS() void {
+    if (active_transaction_stack) |*stack| {
+        stack.deinit();
+        active_transaction_stack = null;
+    }
+    if (active_span_stack) |*stack| {
+        stack.deinit();
+        active_span_stack = null;
+    }
+}
+
+pub fn getActiveTransaction() ?*Transaction {
+    if (active_transaction_stack) |*stack| {
+        if (stack.items.len > 0) {
+            return stack.items[stack.items.len - 1];
+        }
+    }
+    return null;
+}
+
+pub fn getActiveSpan() ?*Span {
+    if (active_span_stack) |*stack| {
+        if (stack.items.len > 0) {
+            return stack.items[stack.items.len - 1];
+        }
+    }
+    return null;
+}
+
+fn shouldSample(client: *const SentryClient, ctx: *const TransactionContext) bool {
+    // 1. Use parent sampling decision if available
+    if (ctx.sampled) |sampled| {
+        std.log.debug("Using parent sampling decision: {?}", .{sampled});
+        return sampled;
+    }
+
+    // 2. Use sample_rate
+    const rate = client.options.sample_rate orelse return false;
+
+    if (rate <= 0.0) {
+        std.log.debug("Dropping transaction: sample_rate is {d}", .{rate});
+        return false;
+    }
+
+    if (rate >= 1.0) {
+        return true;
+    }
+
+    return generateSampleDecision(rate);
+}
+
+fn generateSampleDecision(sample_rate: f64) bool {
+    var prng = std.Random.DefaultPrng.init(@intCast(std.time.microTimestamp()));
+    const random_value = prng.random().float(f64);
+
+    if (random_value < sample_rate) {
+        return true;
+    }
+
+    std.log.debug("Dropping transaction: random value {d} >= sample rate {d}", .{ random_value, sample_rate });
+    return false;
+}
+
+pub fn startTransaction(allocator: Allocator, name: []const u8, op: []const u8) !?*Transaction {
+    const client = scope.getClient() orelse {
+        std.log.debug("No client available, cannot start transaction", .{});
+        return null;
+    };
```

Should this return an error instead of null if it fails because no client? It seems like the function returns `null` if it's sampled